ruby/lib/fluent_tools/converter.rb

```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+require 'open3'
+require 'pathname'
+require 'fileutils'
+
+module FluentTools
+  # Core converter class that interfaces with the fluent-tools Rust binary
+  # Handles file validation, output directory creation, and command execution
+  class Converter
+    def initialize
+      @binary_path = find_binary_path
+    end
+
+    # Convert Fluent file to Android XML
+    def fluent_to_android(input_path, output_path)
+      validate_input_file(input_path)
+      ensure_output_directory(output_path)
+
+      cmd = [@binary_path, 'android', 'from-fluent', '-i', input_path, '-o', output_path]
+      execute_command(cmd)
+    end
+
+    # Convert Android XML to Fluent file
+    def android_to_fluent(input_path, output_path, original_fluent: nil)
+      validate_input_file(input_path)
+      ensure_output_directory(output_path)
+
+      cmd = [@binary_path, 'android', 'to-fluent', '-i', input_path, '-o', output_path]
+      cmd += ['--original-fluent', original_fluent] if original_fluent
+
+      execute_command(cmd)
+    end
+
+    # Convert Fluent file to PO
+    def fluent_to_po(input_path, output_path, locale: 'en-US')
+      validate_input_file(input_path)
+      ensure_output_directory(output_path)
+
+      cmd = [@binary_path, 'po', 'from-fluent', '-i', input_path, '-o', output_path, '-l', locale]
+      execute_command(cmd)
+    end
+
+    # Convert PO file to Fluent
+    def po_to_fluent(input_path, output_path)
+      validate_input_file(input_path)
+      ensure_output_directory(output_path)
+
+      cmd = [@binary_path, 'po', 'to-fluent', '-i', input_path, '-o', output_path]
+      execute_command(cmd)
+    end
+
+    private
+
+    def find_binary_path
+      binary_name = FluentTools::Utils::BINARY_NAME
+
+      # 1. Standard gem installation (works for all gem contexts)
+      gem_binary = File.join(__dir__, '..', '..', 'bin', binary_name)
+      return gem_binary if File.executable?(gem_binary)
+
+      # 2. Development context
+      if development_context?
+        dev_binary = File.join(project_root, 'target', 'release', binary_name)
+        return dev_binary if File.executable?(dev_binary)
+
+        ruby_binary = File.join(project_root, 'ruby', 'bin', binary_name)
+        return ruby_binary if File.executable?(ruby_binary)
+      end
+
+      # 3. System PATH
+      system_binary = `which #{binary_name} 2>/dev/null`.strip
+      return system_binary unless system_binary.empty?
+
+      # If nothing found, return the expected gem path for better error messages
+      gem_binary
+    end
+
+    def development_context?
+      @project_root ||= find_project_root
+      !@project_root.nil?
+    end
+
+    def project_root
+      @project_root ||= find_project_root
+    end
+
+    def find_project_root
+      Pathname.new(__dir__).ascend do |dir|
+        return dir.to_s if File.exist?(File.join(dir, 'Cargo.toml'))
+      end
+      nil
+    end
+
+    def validate_input_file(path)
```

Comment:
I'd suggest to use `!` given it's validation that raises if it doesn't pass
```suggestion
    def validate_input_file!(path)
```

Reply:
Updated on 05d1a515d4134732c79de51d07d485d5f2e63227